src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaJobWorker.cs

```diff
@@ -0,0 +1,73 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Extensions.DependencyInjection;
+using Microsoft.Health.Fhir.Core;
+using Microsoft.Health.Fhir.SqlServer.Configs;
+using Microsoft.Health.Fhir.SqlServer.Features.Schema.Messages.Get;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Schema
+{
+    /// <summary>
+    /// The worker responsible for running the schema job.
+    /// </summary>
+    public class SchemaJobWorker
+    {
+        private readonly Func<IScoped<ISchemaDataStore>> _schemaDataStoreFactory;
+        private readonly SqlServerDataStoreConfiguration _sqlServerDataStoreConfiguration;
+        private readonly ILogger _logger;
+
+        public SchemaJobWorker(Func<IScoped<ISchemaDataStore>> schemaDataStoreFactory, SqlServerDataStoreConfiguration sqlServerDataStoreConfiguration, ILogger<SchemaJobWorker> logger)
+        {
+            EnsureArg.IsNotNull(schemaDataStoreFactory, nameof(schemaDataStoreFactory));
+            EnsureArg.IsNotNull(sqlServerDataStoreConfiguration, nameof(sqlServerDataStoreConfiguration));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _schemaDataStoreFactory = schemaDataStoreFactory;
+            _sqlServerDataStoreConfiguration = sqlServerDataStoreConfiguration;
+            _logger = logger;
+        }
+
+        public async Task ExecuteAsync(SchemaInformation schemaInformation, string instanceName, CancellationToken cancellationToken)
+        {
+            using (IScoped<ISchemaDataStore> store = _schemaDataStoreFactory())
+            {
+                await store.Value.InsertInstanceSchemaInformation(instanceName, schemaInformation, cancellationToken);
+            }
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    _logger.LogInformation($"Polling started at {Clock.UtcNow}");
+
+                    using (IScoped<ISchemaDataStore> store = _schemaDataStoreFactory())
+                    {
+                        GetCompatibilityVersionResponse getCompatibilityVersion = await store.Value.GetLatestCompatibleVersionAsync(cancellationToken);
+
+                        await store.Value.UpsertInstanceSchemaInformation(instanceName, getCompatibilityVersion.Versions, schemaInformation.Current.GetValueOrDefault(), cancellationToken);
```

Comment:
Just for my understanding: why is `InstanceSchema.MaxVersion` and `InstanceSchema.MinVersion` getting set to the compatible versions?

Reply:
Good question. I was working on some refactoring and logic updates yesterday and this is also updated. Now it is set from SchemaInformation property to ensure that if server is restarted and any change in the property should also get updated in the InstanceSchema table.

Reply:
Nice, this makes sense.

Would it be possible to remove the C# call to `GetCurrentSchemaVersion()` and instead modify the SQL logic in the `UpsertInstanceSchema` stored procedure? You might be able to set `InstanceSchema.CurrentVersion` to `MAX(SchemaVersion.Version) WHERE Status = 'complete'` directly.

This would reduce SQL calls and would allow you to get-and-update concurrently.